cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java

```diff
@@ -197,6 +203,54 @@ private void upgradeSystemDatasets(Injector injector) throws Exception {
     }
   }
 
+  private void migrateMetricsData(Injector injector) throws Exception {
+    // find version to migrate
+    ProjectInfo.Version version = findMetricsTableVersion(injector.getInstance(CConfiguration.class),
+                                                  injector.getInstance(Configuration.class));
+    if (version != null) {
+      DatasetFramework framework = createRegisteredDatasetFramework(injector);
+      // migrate metrics data
+      DefaultMetricDatasetFactory.migrateMetricsData(injector.getInstance(CConfiguration.class), framework, version);
+    }
+  }
+
+  private ProjectInfo.Version findMetricsTableVersion(CConfiguration cConf, Configuration hConf) {
+    // Upgrade all datasets in system namespace
+    boolean metricsTable27Found = false;
+    boolean metricsTable26Found = false;
```

Comment:
I'd add a comment with the logic, how we figure out the version. And I think it should be like this:
* figure out what is the latest working version of CDAP (we just talked with @bdmogal about that, the upgrade tool framework will give you any of 2.8.x or "earlier" currently)
* if latest is 2.8.x - nothing to do, if "pre-2.8", proceed to next step
* find a most recent metrics table: start by looking for 2.7, then 2.6 (I'd reformat the code to not look for 2.6 if we found 2.7)
* if you found 2.7 - try upgrade from there, if not - try upgrade from 2.6